### ISSUE TITLE
Allow returning of test results from Trainer.test

### ISSUE THREAD

**Issue by Rizhiy** [PyTorchLightning/pytorch-lightning#7653]
## 🚀 Feature
Currently, `Trainer.test` only returns logged metrics, this is very limiting.

### Motivation

I have a lot of different results which are produced by tests, including figures and sample predictions.
At present, I cannot return them from test since it only allows returning of logged values which it seems must be scalars.

As a workaround I have to separately predict outputs, get targets and then match them together, which is bug-prone.

### Pitch

`Trainer.test` should return output of `Module.test_epoch_end`.

### Alternatives

N/A

### Additional context

N/A

**Comment by Rizhiy** [comment]
I figured another workaround: I can save results to a file in `test_epoch_end` and then load them where I need it, but this is ugly.

**Comment by ananthsub** [comment]
One thing to consider: For large test datasets, caching all step outputs by default will run into out of memory errors. Additionally, near the system memory boundary, performance will be dramatically worse as each allocation will take much more time

**Comment by XiaomoWu** [comment]
@Rizhiy Agree. By the way, when you generate results from `test_epoch_end`, are your results split into `N` parts where N is the number of GPUs? I find `test_epoch_end` only collects results from one GPU.

**Comment by Rizhiy** [comment]
@XiaomoWu You can use `self.all_gather` to get results from different devices onto one.

**Comment by awaelchli** [comment]
Is the [Trainer.predict() API](https://pytorch-lightning.readthedocs.io/en/1.3.1/common/trainer.html#predict) what you are looking for? It will gather your results across all GPUs and by default returns them in a list of dictionaries. 

One can also optionally implement [predict_step](https://pytorch-lightning.readthedocs.io/en/1.3.1/common/lightning_module.html#predict-step) on the LM.

**Comment by Rizhiy** [comment]
@awaelchli No, I need results of test itself, e.g. metric & figures. To calculate that I need access to targets in easy to use manner, so inside `test_*`.

**Comment by versatran01** [comment]
I found this a bit annoying too. I ended up just saving all my metrics to some random member of my LM in test_epoch_end. Something like

```python
def test_epoch_end(self, outputs):
	self.test_metrics = compute_metrics()
```

**Comment by MohammedAljahdali** [comment]
> Is the [Trainer.predict() API](https://pytorch-lightning.readthedocs.io/en/1.3.1/common/trainer.html#predict) what you are looking for? It will gather your results across all GPUs and by default returns them in a list of dictionaries.
> 
> One can also optionally implement [predict_step](https://pytorch-lightning.readthedocs.io/en/1.3.1/common/lightning_module.html#predict-step) on the LM. You could for example move your prediction code from test_step to predict_step. Or you could call test_step() in predict_step()

I have a use case where on the test dataset I would do extra metrics calculations, as well as, logging the prediction and the correct target (They are strings), I would like to be able to return these things from `test_epoch_end` in`trainer.test()`.

Another option, would be to allow `self.log(.... reduce_fx=None)` or `self.log(.... reduce_fx=lambda x: x)`

**Comment by carmocca** [comment]
We cannot provide an automated of gathering whatever the user provides. The closest thing we have is `predict` as described in the above comment: https://github.com/PyTorchLightning/pytorch-lightning/issues/7653#issuecomment-848024353

If we were to find a solution to this problem, it would likely not be from whatever is returned from `test_epoch_end` as overriding that hook means outputs are kept in memory but we cannot know whether this is desired by the user.

Additionally one might want to return data generated from a different hook, so choosing this one could be limiting.

In my opinion, the alternative suggested of saving this data in the model and optionally `all_gather`ing it is fine.

**Comment by dr-upsilon** [comment]
I find the following approach solves my problem, i.e. using a callback to store the data required.

`import pytorch_lightning as pl
import torch
from torch.utils.data import DataLoader, TensorDataset

class CustomTestCallback(pl.Callback):
    def __init__(self):
        super().__init__()
        self.test_outputs = None

    def on_test_end(self, trainer, pl_module):
        # Assuming the test outputs are stored in trainer.test_loop.outputs
        self.test_outputs = torch.cat([x['preds'] for x in trainer.test_loop.outputs])

# Example model
class SimpleModel(pl.LightningModule):
    def __init__(self):
        super(SimpleModel, self).__init__()
        self.layer = torch.nn.Linear(10, 1)

    def forward(self, x):
        return self.layer(x)

    def test_step(self, batch, batch_idx):
        x, y = batch
        preds = self(x)
        return {'preds': preds}

    def configure_optimizers(self):
        return torch.optim.SGD(self.parameters(), lr=0.01)

# Example data
x = torch.randn(100, 10)
y = torch.randn(100, 1)
dataset = TensorDataset(x, y)
dataloader = DataLoader(dataset, batch_size=32)

# Initialize model and trainer
model = SimpleModel()
custom_callback = CustomTestCallback()
trainer = pl.Trainer(callbacks=[custom_callback])

# Run test
trainer.test(model, dataloaders=dataloader)

# Access the test outputs
test_outputs = custom_callback.test_output`


desclaimer: The example code is generated from copilot. It works for me by following the same approach in my project.
